src/lib/Interpreter.hs

```diff
@@ -321,6 +334,18 @@ interpIndexSetSize ixTy = liftBuilderInterp $ indexSetSize $ sink ixTy
 interpIntToIndex :: EnvReader m => IxType n -> Atom n -> m n (Atom n)
 interpIntToIndex ixTy i = liftBuilderInterp $ intToIndex (sink ixTy) (sink i)
 
+-- A variant of `indices` that accepts an expected number of them and
+-- only tries to construct the set if the expected number matches the
+-- given index set's size.
+indicesLimit :: EnvReader m => Int -> IxType n -> m n (Either Int32 [Atom n])
+indicesLimit sizeReq ixTy = do
+  ~(IdxRepVal size) <- interpIndexSetSize ixTy
+  if size == fromIntegral sizeReq then
+    Right <$> forM [0..size-1] \i -> interpIntToIndex ixTy $ IdxRepVal i
```

Comment:
This `forM` will likely run in a virtualized monad that lifts to interpreter at every iteration. Perhaps we could just run it all in the interpreter monad?

Reply:
Done.